### ISSUE TITLE
When "padding_idx" != 1, "negative_targets" incorrect?

### ISSUE THREAD

**Issue by wyshi** [facebookresearch/unlikelihood_training#3]
Thanks for sharing the code!

When running the script on my own data, I encountered the error "RuntimeError: Invalid index in scatter". I figured out the reason to be the following,

https://github.com/facebookresearch/unlikelihood_training/blob/723747171a3fa909cda68df399e39f0a3e5067d9/custom/candidate_penalty_ce_loss.py#L54

If `padding_idx != 1 or -1` (say 10000), line 54 above  will make the upper triangle of `ctx_cands_` all `10000*100000`, which resulted in the invalid index problem. And also, even if `padding_idx*padding_idx` is within valid index (say 10*10=100), it would still cast the wrong index into `negative_targets`.

**A possible solution** could be, adding
                
`ctx_cands = ctx_cands.masked_fill(ctx_cands == (self.padding_idx**2), self.padding_idx)`

after line 58 below
https://github.com/facebookresearch/unlikelihood_training/blob/723747171a3fa909cda68df399e39f0a3e5067d9/custom/candidate_penalty_ce_loss.py#L58

**Comment by uralik** [comment]
this is right, we assumed here that the padding index is 1 (as fairseq uses). In other words, code assumes that padding_idx * padding_idx = padding_idx. Your proposed approach may work, but I didn't test it out myself.
There are many ways to achieve the desired previous context mask as written in the code comment: `# E.g. DABCC | D | EFFGD => {A,B,C} are negative targets.`

Sorry for such delay in response, for some reason I just realized that it is still not answered. I hope it worked out for you in the end! Please feel free to reopen for more questions later!
